src/MongoDB.EntityFrameworkCore/Storage/MongoDatabaseWrapper.cs

```diff
@@ -217,4 +148,148 @@ private static string GetId(IUpdateEntry entry)
             _ => throw new NotSupportedException($"Unknown type for _id property on entity type ${entry.EntityType.DisplayName()}")
         };
     }
+
+    private IEnumerable<MongoUpdate> ConvertUpdateEntriesToMongoUpdates(IEnumerable<IUpdateEntry> entries)
+    {
+        return
+            GetAllChangedRootEntries(entries)
```

Comment:
I'm not 100% sure of `GetAllChangedRootEntries` but for the purposes of this PR I am assuming it is correct.

Example question: if a nested document has a state of `Modified` does that automatically mean the root document also has a state of `Modified`? If not the root document won't get saved.

Reply:
Yes, that's exactly what it does - pushes the change indicator on embedded owned/entities up to the root/owned entity.

Reply:
Is there a bug on this line then?

https://github.com/mongodb/mongo-efcore-provider/blob/main/src/MongoDB.EntityFrameworkCore/Storage/MongoDatabaseWrapper.cs#L121

Seems like it should use `==` instead of `!=`.

Reply:
Question about this line also:

https://github.com/mongodb/mongo-efcore-provider/blob/main/src/MongoDB.EntityFrameworkCore/Storage/MongoDatabaseWrapper.cs#L127

Won't the `root` end up in the `HashSet` twice? Once `Unchanged` and again `Modified`? Similarly, couldn't the `root` end up in the `HashSet` multiple times (all as `Modified`) if the same root entity has more than one modified owned type?

Reply:
A hashset does not allow duplicates so that is fine.

We only want to change the root to modified if it was unmodified. We do not want to change it from added or deleted to modified (which the Cosmos provider does but is a bug afaik)

Reply:
I did notice that it was a HashSet...

So are you assuming that the `IUpdateEntry` implementation of `Equals` uses reference equality?

Are you assuming that changing the state of an `IUpdateEntry` (by changing the `EntityState` property) doesn't affect the results of `Equals` and `GetHashCode`?

Using HashSet with mutable types is perilous.

Reply:
That is true but it is fine here - the Cosmos provider and EF internals use HashSet's of IUpdateEntity and the only implementation has neither Equals nor GetHashCode defined so it is all reference based on the mutable types.